Userland/Libraries/LibWeb/ARIA/RoleType.cpp

```diff
@@ -0,0 +1,282 @@
+{
+    auto state_object = TRY(object.add_object("state"sv));
+    for (auto const state : supported_states()) {
+        auto value = TRY(ARIA::state_or_property_to_string_value(state, m_data, default_value_for_property_or_state(state)));
+        TRY(state_object.add(ARIA::state_or_property_to_string(state), value));
+    }
+    TRY(state_object.finish());
+
+    auto properties_object = TRY(object.add_object("properties"sv));
+    for (auto const property : supported_properties()) {
+        auto value = TRY(ARIA::state_or_property_to_string_value(property, m_data, default_value_for_property_or_state(property)));
+        TRY(properties_object.add(ARIA::state_or_property_to_string(property), value));
+    }
+    TRY(properties_object.finish());
+
+    auto required_states_object = TRY(object.add_object("required_state"sv));
+    for (auto const state : required_states()) {
+        auto value = TRY(ARIA::state_or_property_to_string_value(state, m_data, default_value_for_property_or_state(state)));
+        TRY(required_states_object.add(ARIA::state_or_property_to_string(state), value));
+    }
+    TRY(required_states_object.finish());
+
+    auto required_properties_object = TRY(object.add_object("required_properties"sv));
+    for (auto const property : required_properties()) {
+        auto value = TRY(ARIA::state_or_property_to_string_value(property, m_data, default_value_for_property_or_state(property)));
+        TRY(required_properties_object.add(ARIA::state_or_property_to_string(property), value));
+    }
+    TRY(required_properties_object.finish());
+
+    auto prohibited_states_object = TRY(object.add_object("prohibited_state"sv));
+    for (auto const state : prohibited_states()) {
+        auto value = TRY(ARIA::state_or_property_to_string_value(state, m_data, default_value_for_property_or_state(state)));
+        TRY(prohibited_states_object.add(ARIA::state_or_property_to_string(state), value));
+    }
+    TRY(prohibited_states_object.finish());
+
+    auto prohibited_properties_object = TRY(object.add_object("prohibited_properties"sv));
+    for (auto const property : required_properties()) {
+        auto value = TRY(ARIA::state_or_property_to_string_value(property, m_data, default_value_for_property_or_state(property)));
+        TRY(prohibited_properties_object.add(ARIA::state_or_property_to_string(property), value));
+    }
+    TRY(prohibited_properties_object.finish());
+
+    return {};
+}
+
+ErrorOr<NonnullOwnPtr<RoleType>> RoleType::build_role_object(Role role, bool focusable, AriaData const& data)
+{
+    if (is_abstract_role(role))
+        return Error::from_string_view("Cannot construct a role object for an abstract role."sv);
+
+    switch (role) {
+    case Role::alert:
+        return adopt_nonnull_own_or_enomem(new (nothrow) Alert(data));
+    case Role::alertdialog:
+        return adopt_nonnull_own_or_enomem(static_cast<Alert*>((new (nothrow) AlertDialog(data))));
```

Comment:
I might be missing something, but these `static_cast`s seem unnecessary. `build_role_object` returns a `RoleType` pointer anyway.

Reply:
The casts are necessary to make the compiler happy. Because some roles have multiple parent classes, the compiler seems to have trouble figuring out how an instance of that class is a RoleType, we need to cast to help the compiler along.
![image](https://github.com/SerenityOS/serenity/assets/9423342/fe9b8aad-ed97-4e17-99c9-94ea47ab46c2)

Reply:
Gotcha. I learned a thing! :^)